app/javascript/packs/outdoor_map_builder.js

```diff
@@ -162,6 +162,115 @@ console.log('[MAP] Layers built');
 
 // mapbox token
 
+function allBuildings() {
+  var results = []
+  for (const campus of campusNames) {
+    let buildingsOfCampus = layers[campus];
+    results.push(...buildingsOfCampus.getLayers())
+  }
+  return results
+}
+
+function buildingsByQuery(query) {
+  return allBuildings().filter((buildingLayer) => {
+    // The HPIGeocoder and the highlighting of buildings rely on this fact, which happens to always be the case.
+    // Please inform someone if this warning occurs :)
+    if(buildingLayer.getLayers().length > 1){
+      console.log("WARNING: Not only one inner building layer in:", buildingLayer);
+    }
+    const innerBuildingLayer = buildingLayer.getLayers()[0];
+    const buildingInfo = innerBuildingLayer.feature;
+    return buildingInfo.properties.name && buildingInfo.properties.name.toLowerCase().indexOf(query.toLowerCase()) > -1;
+  })
+}
+
+function buildingAtLocation(location) {
+  // different representation of the position for the pointInPolygon-method
+  const locationArray = [location.lng, location.lat]
+  for(const buildingLayer of allBuildings()) {
+    const innerBuildingLayer = buildingLayer.getLayers()[0];
+    const buildingInfo = innerBuildingLayer.feature
+    const polygonCoordinates = buildingInfo.geometry.coordinates[0];
+    if(pointInPolygon(locationArray, polygonCoordinates)) return buildingLayer
+    // The next line ensures that the buildings are highlighted and properly geocoded,
+    // when they are being routed to. This is because we route to the center of the building.
+    if(innerBuildingLayer.getCenter().toBounds(1).contains(location)) return buildingLayer
```

Comment:
I don't understand this line, isn't the center always inside the building? Or does this account for things like patios?

Reply:
For buildings like the main building and the buildings of the studentenwohnheim it is not the case that the center is inside their bounding polygon.